packages/embarkjs/embarkjs/src/lib/blockchain.js

```diff
@@ -363,6 +374,24 @@ Contract.prototype.send = function(value, unit, _options) {
   return this.blockchainConnector.send(options);
 };
 
+Contract.prototype.deployed = function() {
+  if (this.address === undefined && this._deployPromise === undefined) {
+    const err = Error("contract deployment hasn't happened yet");
+    return Promise.reject(err);
+  }
+
+  // This keeps the original deploy promise so we can fire it as soon as web3 tells us
+  // it's done.
+  if (this._deployPromise) {
+    return this._deployPromise;
+  }
+
+  // Already deployed, just resolve.
+  return Promise.resolve(this);
```

Comment:
Does returning `this` return the web3 contarct or the `EmbarkJS.Contract instance`? It's unclear with all the prototype changes

Reply:
It will be the `EmbarkJS.Contract` instance since it's an instance method in which `this` is a reference to its instance.

Reply:
Ok, I think there is a bug with `deployed` then.
You see, in `deploy` , when the deployment is done, we send a new EmbarkJS.Contract instance with the address set to it, but here, we return the current instance, which will not have an address set, since we do not set the address to our current instance in `deploy`.

So, what I think is that `deployed` should resolve with the new Instance containing the new address. This also makes sense, because it's possible to call `deploy` on the same `EmbarkJS.Contract` instance with different arguments and it would deploy to a different address.

Reply:
> So, what I think is that deployed should resolve with the new Instance containing the new address.

It actually will if `deployPromise` isn't undefined, which means `deploy()` has been called before. In that case the promise will resolve with the new instance instead. But there's another issue, see my comment below.